harness/determined/experimental/_local_execution.py

```diff
@@ -0,0 +1,115 @@
+    """
+    Create a local experiment configuration based on an input configuration and
+    defaults. Use a shallow merging policy to overwrite our default
+    configuration with each entire subconfig specified by a user.
+
+    The defaults and merging logic is not guaranteed to match the logic used by
+    the Determined master. This function also does not do experiment
+    configuration validation, which the Determined master does.
+    """
+
+    input_config = input_config.copy() if input_config else {}
+    config_keys_to_ignore = {
+        "bind_mounts",
+        "checkpoint_storage",
+        "environment",
+        "resources",
+        "optimizations",
+    }
+    for key in config_keys_to_ignore:
+        if key in input_config:
+            logging.info(
+                f"'{key}' configuration key is not supported by local test mode and will be ignored"
+            )
+            del input_config[key]
+
+    return {**constants.DEFAULT_EXP_CFG, **input_config}
+
+
+def _make_local_execution_env(
+    config: Optional[Dict[str, Any]], hparams: Optional[Dict[str, Any]] = None,
+) -> Tuple[det.EnvContext, det.RendezvousInfo, horovod.HorovodContext]:
+    config = det.ExperimentConfig(_make_local_execution_exp_config(config))
+    hparams = hparams or {}
+    use_gpu, container_gpus, slot_ids = _get_gpus()
+    local_rendezvous_ports = (
+        f"{constants.LOCAL_RENDEZVOUS_PORT},{constants.LOCAL_RENDEZVOUS_PORT+1}"
+    )
+
+    env = det.EnvContext(
+        master_addr="",
+        master_port=1,
+        container_id="local_mode",
+        experiment_config=config,
+        hparams=hparams,
+        initial_workload=workload.train_workload(1, 1, 1, config.batches_per_step()),
+        latest_checkpoint=None,
+        use_gpu=use_gpu,
+        container_gpus=container_gpus,
+        slot_ids=slot_ids,
+        debug=config.debug_enabled(),
+        workload_manager_type="",
+        det_rendezvous_ports=local_rendezvous_ports,
+        det_trial_runner_network_interface=constants.AUTO_DETECT_TRIAL_RUNNER_NETWORK_INTERFACE,
+        det_trial_id="1",
+        det_experiment_id="1",
+        det_cluster_id="local_mode",
+        trial_seed=config.experiment_seed(),
+    )
+    rendezvous_ports = env.rendezvous_ports()
+    rendezvous_info = det.RendezvousInfo(
+        addrs=[f"0.0.0.0:{rendezvous_ports[0]}"], addrs2=[f"0.0.0.0:{rendezvous_ports[1]}"], rank=0
+    )
+    hvd_config = horovod.HorovodContext.from_configs(
+        env.experiment_config, rendezvous_info, env.hparams
+    )
+
+    return env, rendezvous_info, hvd_config
+
+
+@contextlib.contextmanager
+def _local_execution_manager(context_dir: pathlib.Path) -> Iterator:
+    """
+    A context manager used for local execution to mimic the environment of trial
+    container execution.
+
+    It does the following things:
+    1. Set the working directory to be the current directory.
+    2. Add the context directory to importing paths.
```

Comment:
What we are actually doing is adding the CWD to the import path, which is why we add `""` and not `str(context_dir)`.

question:  Why is this necessary?

I know we do it in the determined cli because the tooling around the CLI actually forces the local directory to not be on the sys.path, but do we expect that to be the case in user code?

(I guess I could see the argument that it doesn't hurt to have the CWD on the sys.path twice)

Reply:
1 is necessary because there are some places making use of the current working directory to resolve the file path, maybe checkpoints.
2 is necessary because there are some places making use of the importing paths to import files, like loading an implementation.
Don't remember the exact places.

> I know we do it in the determined cli because the tooling around the CLI actually forces the local directory to not be on the sys.path, but do we expect that to be the case in user code?

We only do it for local execution, including local test mode and local inference mode.

> (I guess I could see the argument that it doesn't hurt to have the CWD on the sys.path twice)

We can replace the first element rather than adding a new element. Doesn't seem matter to me.